SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp

```diff
@@ -949,3 +952,124 @@ void USpatialSender::AddTombstoneToEntity(const Worker_EntityId EntityId)
+	{
+		int32_t TmpInitialOnlyComponentId = SubobjectSchemaDataKV.Value.SchemaComponents[SCHEMA_InitialOnly];
+		if (TmpInitialOnlyComponentId && TmpInitialOnlyComponentId != SCHEMA_Invalid)
+		{
+			OutInitialOnlyComponentIds.Add(TmpInitialOnlyComponentId);
+		}
+	}
+
+	// Process InitialOnly data of dynamic subobject
+	for (auto ActorComponent : ActorComponents)
+	{
+		FString TmpClassPathName = ActorComponent->GetClass()->GetPathName();
+		if (SchemaDataBase->SubobjectClassPathToSchema.Contains(TmpClassPathName))
+		{
+			FSubobjectSchemaData& SubobjectSchemaData = SchemaDataBase->SubobjectClassPathToSchema[TmpClassPathName];
+			for (auto DynamicSubobjectComponents : SubobjectSchemaData.DynamicSubobjectComponents)
+			{
+				int32_t TmpInitialOnlyComponentId = DynamicSubobjectComponents.SchemaComponents[SCHEMA_InitialOnly];
+				if (TmpInitialOnlyComponentId && TmpInitialOnlyComponentId != SCHEMA_Invalid)
+				{
+					OutInitialOnlyComponentIds.Add(TmpInitialOnlyComponentId);
+				}
+			}
+		}
+	}
+}
+
+void USpatialSender::SendInitialOnlyRequest(Worker_EntityId EntityId)
+{
+	TArray<Worker_ComponentId> LocalInitialOnlyComponentIds;
+	GetInitialOnlyComponentIds(EntityId, LocalInitialOnlyComponentIds);
+
+	if (0 == LocalInitialOnlyComponentIds.Num())
+	{
+		return;
+	}
+
+	InitialOnlyRequests.Add(TPair<Worker_EntityId, TSet<Worker_ComponentId>>(EntityId, LocalInitialOnlyComponentIds));
+}
+
+void USpatialSender::CancelInitialOnlyRequest(Worker_EntityId EntityId)
+{
+	InitialOnlyRequests.Remove(EntityId);
+}
+
+void USpatialSender::FlushInitialOnlyRequests()
+{
+	if (0 == InitialOnlyRequests.Num())
+	{
+		return;
+	}
+
+	int32_t EntityCount = 0;
+
+	TSet<Worker_ComponentId> ComponentIdsSet;
+	TArray<Worker_Constraint> EntityConstraintArray;
+	TArray<Worker_EntityId> RemovedEntities;
+
+	for (auto Entity : InitialOnlyRequests)
+	{
+		Worker_EntityId TmpEntityId = Entity.Key;
+		Worker_Constraint Constraints;
+		Constraints.constraint_type = WORKER_CONSTRAINT_TYPE_ENTITY_ID;
+		Constraints.constraint.entity_id_constraint.entity_id = TmpEntityId;
+
+		EntityConstraintArray.Add(Constraints);
+		ComponentIdsSet.Append(Entity.Value);
+		RemovedEntities.Add(Entity.Key);
+
+		if (++EntityCount >= MaxInitialOnlyRequestPerTick)
+		{
+			break;
+		}
+	}
+
+	for (auto RemovedEntityId : RemovedEntities)
+	{
+		InitialOnlyRequests.Remove(RemovedEntityId);
+	}
+
+	Worker_Constraint InitialOnlyConstraints;
+	InitialOnlyConstraints.constraint_type = WORKER_CONSTRAINT_TYPE_OR;
+	InitialOnlyConstraints.constraint.and_constraint.constraint_count = EntityConstraintArray.Num();
+	InitialOnlyConstraints.constraint.and_constraint.constraints = EntityConstraintArray.GetData();
+
+	const TArray<Worker_ComponentId>& TmpComponentIds = ComponentIdsSet.Array();
+
+	Worker_EntityQuery InitialOnlyQuery{};
+	InitialOnlyQuery.constraint = InitialOnlyConstraints;
+	InitialOnlyQuery.snapshot_result_type_component_id_count = TmpComponentIds.Num();
```

Comment:
Hmm this makes me think we can just generate a component set for all initial only data, then we don't need to worry about preparing which components we're interested in at runtime as we'll know it statically

Reply:
Looks like there's already a component set generated 🎉

Reply:
Yep, I initially wanted to make use of  ComponentSet to request all the initial only data, but @samiwh mentioned once that ComponentSet might not be worked for now, if it does work now so I think it's better to replace with it